pkg/kvstore/consul.go

```diff
@@ -407,3 +453,135 @@ func (c *ConsulClient) DeleteTree(path string) error {
 	_, err := c.Client.KV().DeleteTree(path, nil)
 	return err
 }
+
+// Set sets value of key
+func (c *ConsulClient) Set(key string, value []byte) error {
+	_, err := c.KV().Put(&consulAPI.KVPair{Key: key, Value: value}, nil)
+	return err
+}
+
+// Delete deletes a key
+func (c *ConsulClient) Delete(key string) error {
+	_, err := c.KV().Delete(key, nil)
+	return err
+}
+
+// Get returns value of key
+func (c *ConsulClient) Get(key string) ([]byte, error) {
+	pair, _, err := c.KV().Get(key, nil)
+	if err != nil {
+		return nil, err
+	}
+	if pair == nil {
+		return nil, nil
+	}
+	return pair.Value, nil
+}
+
+// GetPrefix returns the first key which matches the prefix
+func (c *ConsulClient) GetPrefix(prefix string) ([]byte, error) {
+	pairs, _, err := c.KV().List(prefix, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(pairs) == 0 {
+		return nil, nil
+	}
+
+	return pairs[0].Value, nil
+}
+
+// CreateOnly creates a key with the value and will fail if the key already exists
+func (c *ConsulClient) CreateOnly(key string, value []byte, lease bool) error {
+	k := &consulAPI.KVPair{
+		Key:         key,
+		Value:       value,
+		CreateIndex: 0,
+	}
+
+	if lease {
+		id, ok := leaseInstance.(string)
+		if !ok {
+			return fmt.Errorf("argument not a LeaseID")
+		}
+
+		k.Session = id
+	}
+
+	success, _, err := c.KV().CAS(k, nil)
+	if !success || err != nil {
+		return fmt.Errorf("create was unsuccessful: %s", err)
+	}
+
+	return nil
+}
+
+// CreateIfExists creates a key with the value only if key condKey exists
+func (c *ConsulClient) CreateIfExists(condKey, key string, value []byte, lease bool) error {
+	if err := c.CreateOnly(key, value, lease); err != nil {
+		return err
+	}
+
+	// Consul does not support transactions which would allow to check for
+	// the presence of another key
+	masterKey, err := c.Get(condKey)
+	if err != nil || masterKey == nil {
+		c.Delete(key)
+		return fmt.Errorf("conditional key not present")
+	}
+
+	return nil
+}
+
+// ListPrefix returns a map of matching keys
+func (c *ConsulClient) ListPrefix(prefix string) (KeyValuePairs, error) {
+	pairs, _, err := c.KV().List(prefix, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	p := KeyValuePairs{}
```

Comment:
You can allocate a map of the right size from the start:

    p := KeyValuePairs(make(map[string][]byte, len(pairs)))